datastore/google/cloud/datastore/batch.py

```diff
@@ -199,6 +199,28 @@ def put(self, entity):
 
         _assign_entity_to_pb(entity_pb, entity)
 
+    def put_entity_pb(self, entity_pb):
+        if self._status != self._IN_PROGRESS:
+            raise ValueError("Batch must be in progress to put()")
+
+        if entity_pb.key is None:
+            raise ValueError("Entity must have a key")
+
+        if self.project != entity_pb.key.partition_id.project_id:
+            raise ValueError("Key must be from same project as batch")
+
+        # Key is considered partial if neither "id" nor "name" values of the key
+        # "path" field are set
+        is_key_partial = (not entity_pb.key.path or (not entity_pb.key.path[0].name and not
+                          entity_pb.key.path[0].id))
+
+        if is_key_partial:
+            insert_entity_pb = self._add_partial_key_entity_pb()
+            insert_entity_pb.CopyFrom(entity_pb)
```

Comment:
@tseaver 

To make sure I understand you correctly - you mean making sure ``self._partial_key_entities`` is also updated when ``put_entity_pb`` is used or to handle that similarly in here https://github.com/googleapis/google-cloud-python/blob/42c2b28356249cde348eb75780452b97a47d0684/datastore/google/cloud/datastore/batch.py#L278?

One option would be to have two lists - one which stores Python entity objects and the other one which stores EntityPB objects.

This way inside ``_commit`` we would iterate over both and assign keys on both type ob objects. I think this would be the most reasonable approach (other approaches would involve unnecessary conversions back and forth between Entity PB and Entity and this kinda defeats the purpose of the whole change).

Are you fine with that approach?

Reply:
EDIT: It shows that it has been a while since I last worked on that change. I actually already implemented that in the past :)

Having said that, I can still push a change which makes it utilize two lists instead of one to make it a bit more cleaner. And I can also have a look if I can remove the unnecessary conversion between Entity and Entity PB in that commit function.

I will also go ahead and sync my branch with latest master.

Please let me know if there is anything else I'm missing with regards to ``_commit`` or anything else I need to do.

Reply:
@tseaver 

I synced my branch with latest master and pushed an improvement to avoid unnecessary conversion between Entity and Entity PB object - 0a9f2f3705ffe61a25b6680ff8a49fec313a9d36.

While working on this change, I also noticed that the code for determining if Entity Protobuf key is partial wasn't fully correct for hierarchical keys, so I also pushed a fix for that and added multiple test cases which cover scenario for hierarchical keys.

In the end, I also decided to stick with a single list approach. Two lists approach would be more complicated and error prone since it means I would need to track mutations indexes / order for each list and handle them correctly inside ``commit`` function.

If I'm missing something else or there is something else I need to do, please let me know.